### PR TITLE
[Perf] Reduce H2D pageable memory copies

### PR DESCRIPTION
**Pull request by jackcfwang**
## Purpose
Fix H2D (Host-to-Device) pageable memory copy bottleneck in  attention Triton kernel, which caused pipeline bubbles between transformer blocks.

**Root Cause**: `mm_prefix_range_tensor()` recomputes and transfers data from CPU to GPU on every attention call, generating redundant H2D pageable copies.

**Solution**: Cache the computed `mm_prefix_range_tensor` result after first computation to eliminate repeated H2D transfers.

**Performance Impact**: Reduces bubble time between transformer layers, improving end-to-end inference throughput for multimodal models using TRITON_ATTN.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/attention/backends/triton_attn.py [hunk 1]
```diff
@@ -105,15 +113,18 @@ def mm_prefix_range_tensor(self) -> torch.Tensor | None:
...
+        max_ranges = max(len(r) for r in range_lists)
+        # Pad all sequences to the same number of ranges
+        padded = []
+        for r in range_lists:
+            padded_r = list(r) + [(0, 0)] * (max_ranges - len(r))
+            padded.append(padded_r)
+        # Create a single CPU tensor for efficient H2D transfer, avoiding multiple transfers and nested padding operations
+        return torch.tensor(
+            padded, dtype=torch.int32, device=device
+        ).view(num_seqs, max_ranges, 2)
```
[the remark this answers is not in the document]
Comment (JackWang2120) [reply vllm/v1/attention/backends/triton_attn.py]: If we do this, will it generate a res tensor that is too large on the CPU side, affecting the memory space on the CPU side?

[the remark this answers is not in the document]
Comment (jackcfwang) [reply vllm/v1/attention/backends/triton_attn.py]: When I used the internal pressure testing platform to test the code you suggested, with the batch size ranging from 3 to 35, the data obtained indicated that this optimization was slightly less effective than my original implementation.

File: vllm/v1/attention/backends/triton_attn.py [hunk 2]
```diff
@@ -80,16 +80,24 @@ class TritonAttentionMetadata:
...
+        does not trigger redundant H2D / D2D copies.
         """
-        # TODO(Isotr0py): Move to model runner's attention metadata
-        # preparation to avoid duplicate computation.
+        if self._mm_prefix_range_tensor_cache is not None:
+            return self._mm_prefix_range_tensor_cache
+
+        result = self._compute_mm_prefix_range_tensor()
+        self._mm_prefix_range_tensor_cache = result
+        return result
```
Comment (Isotr0py) [inline vllm/v1/attention/backends/triton_attn.py]: I remember that `mm_prefix_range` will probably change after scheduler step (old request finished / new request inserted).

But seems that `_mm_prefix_range_tensor_cache` won't be updated in this case?

Comment (jackcfwang) [reply vllm/v1/attention/backends/triton_attn.py]: @Isotr0py Thank you for your suggestion. I did indeed overlook this case. Currently, I have overridden the `__setattr__` method. When `mm_prefix_range` changes, I set `_mm_prefix_range_tensor_cache` to `None`. Please review this modification and let me know if it complies with the open-source standards.

Comment (jackcfwang) [reply vllm/v1/attention/backends/triton_attn.py]: @Isotr0py I'm not sure if there is a more optimized way to implement this. At present, I can only think of this approach to avoid the cache invalidation issue.

Comment (Isotr0py) [reply vllm/v1/attention/backends/triton_attn.py]: Hmmm, overrides `__setattr__` maybe a little bit hacky. How about hashing the `mm_prefix_range` to match the cached tensor? So when the previous hash doesn't match the new ones, we can know that the cache should be updated.

But I have no idea that whether the hashing will introduce extra overheads tbh.

You can refer to the `HashableDict` and `HashableList` implementation for `mm_prefix_range` since pythonic dict and list are unhashable:
https://github.com/vllm-project/vllm/blob/46f02e00f2d99944ecd24000bcc8be2a9504eafe/vllm/transformers_utils/processor.py#L91-L108

Comment (jackcfwang) [reply vllm/v1/attention/backends/triton_attn.py]: @Isotr0py Thank you for your proposal on using hash mm_prefix_range to determine whether the cache needs to be updated. I wrote a version of the code based on your idea, but on our internal stress testing platform, we found that the **hash** solution performed worse than the **__setattr__** solution. Moreover, as the batch size increased, the overhead brought by hash grew larger and larger, leading to a more severe performance decline.

Comment (Isotr0py) [reply vllm/v1/attention/backends/triton_attn.py]: >  I wrote a version of the code based on your idea, but on our internal stress testing platform, we found that the hash solution performed worse than the setattr solution. Moreover, as the batch size increased, the overhead brought by hash grew larger and larger, leading to a more severe performance decline.

I see, thanks for benchmarking this! If we revert to the method that sets `mm_prefix_range_tensor` together with `mm_prefix_range`, I prefer to set it explicitly at attn metadata preparation stage in model runner instead of overriding `__setattr__`, then we still only have to compute tensor once for each request:
https://github.com/vllm-project/vllm/blob/140cbb118698b0f685f2a1ce2c348962636755c6/vllm/v1/worker/gpu_model_runner.py#L2328-L2334

It will be something looks like this:
```diff
 if isinstance(attn_metadata, list): 
     for ub_metadata in attn_metadata: 
         for _metadata in ub_metadata.values(): 
             _metadata.mm_prefix_range = req_doc_ranges  # type: ignore[attr-defined] 
+++          if hasattr(_metadata, "mm_prefix_range_tensor"):
+++              _metadata.mm_prefix_range_tensor = ...
 else: 
     for _metadata in attn_metadata.values(): 
         _metadata.mm_prefix_range = req_doc_ranges  # type: ignore[attr-defined]
+++      if hasattr(_metadata, "mm_prefix_range_tensor"):
+++          _metadata.mm_prefix_range_tensor = ...
```

WDYT?

**Review by Isotr0py — APPROVED** [review]
LGTM, thanks for your patience!
